tests/mtd_raw/main.c

```diff
@@ -0,0 +1,468 @@
+{
+    bool oob;
+    mtd_dev_t *dev = _get_dev(argc, argv, &oob);
+    uint32_t addr, len;
+
+    if (argc < 4) {
+        printf("usage: %s <dev> <addr> <data>\n", argv[0]);
+        return -1;
+    }
+
+    if (oob) {
+        return -1;
+    }
+
+    addr = atoi(argv[2]);
+    len  = strlen(argv[3]);
+
+    int res = mtd_write(dev, argv[3], addr, len);
+
+    if (res) {
+        printf("error: %i\n", res);
+    }
+
+    return res;
+}
+
+static int cmd_write_page(int argc, char **argv)
+{
+    bool oob;
+    mtd_dev_t *dev = _get_dev(argc, argv, &oob);
+    uint32_t page, offset, len;
+
+    if (argc < 5) {
+        printf("usage: %s <dev> <page> <offset> <len>\n", argv[0]);
+        return -1;
+    }
+
+    if (oob) {
+        return -1;
+    }
+
+    page   = atoi(argv[2]);
+    offset = atoi(argv[3]);
+    len    = strlen(argv[4]);
+
+    int res = mtd_write_page(dev, argv[4], page, offset, len);
+
+    if (res) {
+        printf("error: %i\n", res);
+    }
+
+    return res;
+}
+
+static int cmd_erase(int argc, char **argv)
+{
+    bool oob;
+    mtd_dev_t *dev = _get_dev(argc, argv, &oob);
+    uint32_t addr;
+    uint32_t len;
+
+    if (argc < 4) {
+        printf("usage: %s <dev> <addr> <len>\n", argv[0]);
+        return -1;
+    }
+
+    if (oob) {
+        return -1;
+    }
+
+    addr = atoi(argv[2]);
+    len  = atoi(argv[3]);
+
+    int res = mtd_erase(dev, addr, len);
+
+    if (res) {
+        printf("error: %i\n", res);
+    }
+
+    return res;
+}
+
+static int cmd_erase_sector(int argc, char **argv)
+{
+    bool oob;
+    mtd_dev_t *dev = _get_dev(argc, argv, &oob);
+    uint32_t sector, count = 1;
+
+    if (argc < 3) {
+        printf("usage: %s <dev> <sector> [count]\n", argv[0]);
+        return -1;
+    }
+
+    if (oob) {
+        return -1;
+    }
+
+    sector = atoi(argv[2]);
+
+    if (argc > 3) {
+        count = atoi(argv[3]);
+    }
+
+    int res = mtd_erase_sector(dev, sector, count);
+
+    if (res) {
+        printf("error: %i\n", res);
+    }
+
+    return res;
+}
+
+static void _print_info(mtd_dev_t *dev)
+{
+    printf("sectors: %"PRIu32"\n", dev->sector_count);
+    printf("pages per sector: %"PRIu32"\n", dev->pages_per_sector);
+    printf("page size: %"PRIu32"\n", dev->page_size);
+    printf("total: %lu\n", (unsigned long)_get_size(dev));
+}
+
+static int cmd_info(int argc, char **argv)
+{
+    if (argc < 2) {
+        printf("mtd devices: %d\n", MTD_NUMOF);
+
+        for (int i = 0; i < MTD_NUMOF; ++i) {
+            printf(" -=[ MTD_%d ]=-\n", i);
+            _print_info(_get_mtd_dev(i));
+        }
+        return 0;
+    }
+
+    bool oob;
+    mtd_dev_t *dev = _get_dev(argc, argv, &oob);
+
+    if (oob) {
+        return -1;
+    }
+
+    _print_info(dev);
+
+    return 0;
+}
+
+static int cmd_power(int argc, char **argv)
+{
+    bool oob;
+    mtd_dev_t *dev = _get_dev(argc, argv, &oob);
+    enum mtd_power_state state;
+
+    if (argc < 3) {
+        goto error;
```

Comment:
How about adding


```C
static int _print_usage(const char *progname) {
    printf("usage: %s <dev> <on|off>\n", progname);
    return -1;
}
```

and the replacing all `goto error;` by `return _print_usage(argv[0]);`?

Reply:
If you think that's nicer - I thought for such a simple function it would be better to have it all contained insude one function.

Reply:
But not unlocking the [Go to Hell Achievement](https://www.golem.de/1201/89181.html) is also worth something, if you ask me ;-) (No, I'm not using visual studio ;-))

Reply:
Eh I thought `goto error` was a common pattern.
Anyway, doesn't make much of a difference.